SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -1,41 +1,56 @@
 import Mathlib.RingTheory.Polynomial.Quotient
+import Mathlib.RingTheory.Ideal.Quotient
 import Mathlib.Data.Zmod.Basic
 import SSA.Core.WellTypedFramework
+-- https://github.com/google/heir/blob/a18d4e8ddc0031e2a0e6dd2dd0d7fe289b9d1651/include/Dialect/Poly/IR/PolyDialect.td
+-- https://github.com/j2kun/heir/blob/5c5c0e2ff2ae37a7d1ec5791ec6c38046c4115c1/include/Dialect/Poly/IR/PolyOps.td
+-- https://github.com/google/heir/blob/a18d4e8ddc0031e2a0e6dd2dd0d7fe289b9d1651/tests/poly/poly_ops.mlir
 
 open Polynomial -- for R[X] notation
 
+
 -- Basics from Junfeng Fan and Frederik Vercauteren, Somewhat Practical Fully Homomorphic Encryption
 -- https://eprint.iacr.org/2012/144
-
 variable (q t : Nat) [Fact (q > 1)] (n : Nat)
 
--- TODO: can we make this computable?
--- failed to compile definition, consider marking it as 'noncomputable' because it depends on 'Polynomial.add'', and it does not have executable code
+-- Can we make this computable?
+-- see: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Way.20to.20recover.20computability.3F/near/322382109
+-- and :https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Groebner.20bases
 
 -- Question: Can we make something like d := 2^n work as a macro?
+--
+
 noncomputable def f : (ZMod q)[X] := X^(2^n) + 1
 abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
-
 -- Coefficients of `a : R' q n` are `a\_i : Zmod q`.
 -- TODO: get this from mathlib
-def R.coeff (a : R q n) (i : Nat) : ZMod q := sorry
 
+-- Canonical epimorphism `ZMod q[X] ->*+ R q n`
+abbrev R.fromPoly {q n : Nat} : (ZMod q)[X] → R q n := Ideal.Quotient.mk (Ideal.span {f q n})
 
+noncomputable abbrev R.zero {q n : Nat} : R q n := R.fromPoly 0
+noncomputable abbrev R.one {q n : Nat} : R q n := R.fromPoly 1
+
+noncomputable instance {q n : Nat} : Zero (R q n) := ⟨R.zero⟩
+noncomputable instance {q n : Nat} : One (R q n) := ⟨R.one⟩
+
+axiom R.representative {q n} : R q n → (ZMod q)[X]
```

Comment:
@goens Could you write down a stanza about what this is supposed to be doing? I would also imagine we need a law, that says that `R.fromPoly . R.representative = id`. (note that the other sided inverse condition --- `R.representative . R.fromPoly = id` does not hold).

Reply:
yeah, that's commented in the next line... (because of the same issue with the implicit variable) I'll uncomment it with @ChrisHughes24's fix.